npc-utils.js

```diff
@@ -0,0 +1,551 @@
+
+import * as THREE from 'three';
+import {
+  rootScene,
+} from './renderer.js';
+import physicsManager from './physics-manager.js';
+
+const identityQuaternion = new THREE.Quaternion();
+
+const localVector = new THREE.Vector3();
+const localVoxel = new THREE.Object3D();
+
+const colorIdle = new THREE.Color('rgb(221,213,213)');
+const colorReached = new THREE.Color('rgb(171,163,163)');
+const colorFrontier = new THREE.Color('rgb(92,133,214)');
+const colorStart = new THREE.Color('rgb(0,255,255)');
+const colorDest = new THREE.Color('rgb(255,255,0)');
+const colorPath = new THREE.Color('rgb(149,64,191)');
+const colorPathSimplified = new THREE.Color('rgb(69,0,98)');
+
+class PathFinder {
+  constructor({voxelHeight = 1.5, heightTolerance = 0.6, detectStep = 0.1, maxIterdetect = 1000, maxIterStep = 1000, maxVoxelCacheLen = 10000, debugRender = false}) {
+    /* args:
+      voxelHeight: Voxel height ( Y axis ) for collide detection, usually equal to npc's physical capsule height. X/Z axes sizes are hard-coded 1 now.
+      heightTolerance: Used to check whether currentVoxel can go above to neighbor voxels.
+      detectStep: How height every detecting step moving.
+      maxIterdetect: How many steps can one voxel detecing iterate.
```

Comment:
I still don't understand `maxIterdetect`. Why is it needed? Isn't the max already known from `maxIterStep`?

Reply:
Every voxel need use `overlapBox()` to recursively `detect()` to decide it's `position.y`, `maxIterdetect` is used to limit this recursion.
It's very different from `maxIterStep` which is used to limit A* algorithm's `step()` recursion.